app/main/register-ledger-listeners.ts

```diff
@@ -0,0 +1,151 @@
+import { ipcMain } from 'electron';
+import { BehaviorSubject, Subject, timer, combineLatest, from, of } from 'rxjs';
+import { delay, filter, map, switchMap, take } from 'rxjs/operators';
+import type Transport from '@ledgerhq/hw-transport';
+import TransportNodeHid from '@ledgerhq/hw-transport-node-hid';
+import { safeAwait } from '@blockstack/ui';
+import StacksApp, { LedgerError, ResponseVersion } from '@zondax/ledger-blockstack';
+import {
+  ledgerRequestSignTx,
+  ledgerRequestStxAddress,
+  ledgerShowStxAddress,
+} from './ledger-actions';
+
+type LedgerEvents =
+  | 'create-listener'
+  | 'remove-listener'
+  | 'waiting-transport'
+  | 'disconnected'
+  | 'has-transport';
+
+type LedgerStateEvents =
+  | Record<'name', LedgerEvents>
+  | (Record<'name', keyof typeof LedgerError> & ResponseVersion & { action: 'get-version' });
+
+interface LedgerMessageType {
+  type: 'ledger-event';
+}
+
+export type LedgerMessageEvents = LedgerStateEvents & LedgerMessageType;
+
+const POLL_LEDGER_INTERVAL = 1_250;
+const SAFE_ASSUME_REAL_DEVICE_DISCONNECT_TIME = 2_000;
+
+const ledgerState$ = new Subject<LedgerStateEvents>();
+const ledgerDeviceBusy$ = new BehaviorSubject(false);
+const listeningForDevice$ = new BehaviorSubject(false);
+const checkDisconnect$ = new Subject<void>();
+
+const transport$ = new BehaviorSubject<Transport | null>(null);
+
+export function registerLedgerListeners(webContent: Electron.WebContents) {
+  ledgerState$
+    .pipe(map(ledgerEvent => ({ type: 'ledger-event', ...ledgerEvent })))
+    .subscribe(event => webContent.send('message-event', event));
+}
+
+let subscription: null | ReturnType<typeof TransportNodeHid.listen> = null;
+function createDeviceListener() {
+  if (subscription) subscription.unsubscribe();
+  subscription = TransportNodeHid.listen({
+    next: async event => {
+      if (event.type === 'add') {
+        ledgerState$.next({ name: 'waiting-transport' });
+        if (subscription) subscription.unsubscribe();
+        const [error, ledgerTransport] = await safeAwait(TransportNodeHid.open(event.descriptor));
+        ledgerState$.next({ name: 'has-transport' });
+        if (ledgerTransport) transport$.next(ledgerTransport);
+        if (error) console.log({ error });
+      }
+    },
+    error: e => console.log('err', e),
+    complete: () => console.log('complete'),
+  });
+}
+
+listeningForDevice$.subscribe(listening => {
+  if (listening) {
+    createDeviceListener();
+    return;
+  }
+  subscription?.unsubscribe();
+  const transport = transport$.getValue();
+  if (transport) void transport.close();
+});
+
+const shouldPoll$ = combineLatest([listeningForDevice$, ledgerDeviceBusy$]).pipe(
+  map(([listeningForDevice, deviceBusy]) => Boolean(listeningForDevice && !deviceBusy))
+);
+
+const devicePoll$ = timer(0, POLL_LEDGER_INTERVAL).pipe(
```

Comment:
A comment for this chunk would be useful, like what conditions cause this timer to be created, why a timer is needed, how long it's alive for